feat: implement getServerFunctionMeta

katywings opened the pull request:
This adds a new public `getServerFunctionMeta` API that can be used in "use server" functions to retrieve function related information. It can be used by the enduser as a stable key for server function related caches and debugging.

## Tasks

- [x] Implement basic functionality
- [x] Figure out why importing `getRequestEvent` in `@solidjs/start/server` breaks the vite dev server
- [x] Discuss and eventually change the naming of `getRpcInfo` / `RpcInfo`
    - Renamed to `getServerFunctionMeta / ServerFunctionMeta` 
- [x] Write docs
- [ ] Write tests - (if there is a place to write tests ;))

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Related discussion: nksaraf/vinxi#51

brenelz commented on packages/start/server/rpc.tsx:
```diff
@@ -0,0 +1,8 @@
+  const event = getRequestEvent();
+
+  return event?.context.rpcInfo;
+}
```
maybe this could just be done in user land? not sure

katywings replied:
Yes technically speaking it absolutely could 🙂. The reason why I added this little helper is because the event context type does not have an rpcInfo and I am not sure if/how we can extend the type. Having a helper seems like the easiest way to document the existence of rpcInfo in solid-start 🤔

katywings replied:
(https://github.com/ryansolid/dom-expressions/blob/0aeca28611b5e9d85b135efbfc39892672efac95/packages/dom-expressions/src/server.js#L570)

katywings commented on packages/start/server/rpc.tsx:
```diff
@@ -0,0 +1,8 @@
+export function getRpcInfo(): RpcInfo | undefined {
+  const event = getRequestEvent();
+
+  return event?.context.rpcInfo;
```
I actually thougt about this, but since getRequestEvent doesn't throw an error if no event exists, it seemed more consistent to also throw no error here. Currently if the function is used in a non-server function or outside of a request, in both cases its just returning undefined, which users easily can check with an if condition, if they need to.

katywings replied:
(https://github.com/ryansolid/dom-expressions/blob/0aeca28611b5e9d85b135efbfc39892672efac95/packages/dom-expressions/src/server.js#L570)

katywings replied:
@ryansolid I would also be curious in your opinion on this topic 🙂

ryansolid replied:
I suppose it would be safe for us to throw an error on the server... the reason I didn't was that I could picture libraries using this in situations where people didn't opt into ALS, and also places where they call it client side.. And in those cases you would need to null check and you probably wouldn't want a thrown error.

katywings replied:
I am on the no-throw "side" because it feels like an early assumption at this point - the advantage of not throwing is that the user also can use the `getServerFunctionMeta` to figure out if they are in a server function, which could be useful for helper libraries, eg:

```ts
const someHelper = function(event) {
  if (getServerFunctionMeta()) {
    // Do something special if this helper is called in a server function
  }
}
```

---

That said, I am open to add the throw of course, if thats what you both prefer? If we add a throw to `getServerFunctionMeta`, how about we also add a non-throwing `isServerFunction` helper?

```ts
export const isServerFunction = () => !!getRequestEvent()?.context.serverFunctionMeta
```

katywings replied:
Resolving this for now after a quick chat with Ryan.

nksaraf commented:
since this is related to server functions, lets use some terminology that matches that. `event.serverFunction` maybe? And the helper could be `getServerFunctionEvent` which is a wrapper around `getRequestEvent`. 

But I am still not sure we should expose `name` and `id` separately. These are implementation details. If its a stable key that's needed. I think we should only have `id` maybe which should be opaque. No guarantees about it being same in dev and prod, etc. Just that it'll be stable for one version of the code in that build.

katywings commented:
@nksaraf Would you construct the id how I did it earlier? (`id + "#" + name`)

Do you think this serverFunctionEvent object could include other stuff in the future? If we only return the id, how about we just call the helper `getServerFunctionId`?

Alternatively if you have ideas/plans to include other stuff, how about `getServerFunctionMeta`? I think `Meta` better communicates that the information does not change on each request. `Event` sounds like a thing that changes over time.

nksaraf commented:
> Alternatively if you have ideas/plans to include other stuff, how about `getServerFunctionMeta`? I think `Meta` better communicates that the information does not change on each request. `Event` sounds like a thing that changes over time.

Yup might be better to go with Meta (and just expose the id in there). The id using `id#name` is perfect I think. In production usually we are going to has the `id` part anyway, so its fine.

katywings commented:
@nksaraf I renamed it to `getServerFunctionMeta` and changed the id. I also just tested again the issue with the `getRequestEvent` with `vinxi v0.0.64` . Importing `getRequestEvent` in the `start/server` entry still crashes the dev server with  the following error:
```
node:internal/process/promises:289
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "true".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
```

Do you have an idea why this error is happening and how it can be prevented? I saw that `start/server/spa/StartServer.tsx` also imports and uses `getRequestEvent` without an error. Why does it work there, but fails in my `start/server/serverFunction.tsx` file?

katywings commented:
@nksaraf Any ideas why the getRequestEvent import in solid-start breaks the dev server 😅? This is the only reason why this PR is hanging in "WIP" 🙈

katywings commented:
Retested this with all latest versions. The dev server crash from importing `getRequestEvent` in solid-start internals has been resolved. My random guess is that the error might have been related to c586f14b1a126b06ab57157bf620deab687ddea8.

Anyhow, this PR is ready for further discussion now :).

ryansolid commented:
My only thought here is event.context is an `H3` thing not a start thing so I'd consider using a different location. I haven't really standardized that though.. there is `event.locals`.

I guess we need to make a decision here soon. Right now it is my undocumented playground, but we need to do better.

katywings commented:
> event.context is an H3 thing ... there is event.locals. ...

Okay that is weird, Vscode is showing me h3 types for `getRequestEvent` in my local solid-start workspace, so I assumed that internally in solid-start getRequestEvent directly returns the h3 event - I didn't see that `event.locals` is a thing 😅.

![Bildschirmfoto vom 2024-01-12 12-46-06](https://github.com/solidjs/solid-start/assets/4012401/b4e81640-d6a2-4981-8a49-06f143af1515)

pi0 commented:
Yes, that's precisely the purpose of `event.context` to be extended (and typed) by the frameworks which I wish solid adopted in first place instead of making more wrappers...

katywings commented:
@ryansolid How do we move forward with this PR? 😅

katywings commented:
Gotcha, I am looking forward what you will come up with 🙏. Just regarding "conventions": the new solid start beta changed a lot of things, removed features, removed whole apis, added others, some users have to migrate a lot of code (such as myself), I dont grasp why suddenly with the topic of event locals we should not rethink old conventions. I mean, who if not you, the CEO of signals, is not constantly reworking, rethinking, evolving old conventions 😉. I get that this goes beyond SolidStart 👍.

katywings commented:
Fixed merge conflicts and updated the feature to the current event structure.

ryansolid commented:
It was easier to cherry pick the commit than merge the merge of this taking so long. I will close this. But it is getting in.

katywings commented:
> It was easier to cherry pick

Yeah that makes absolutely sense 😅👍.

> But it is getting in.

Thats some awesome news, thank you!
